File: leet_GRDY_0121_max_profit.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) < 2:
            return 0
        lowest = prices[0]
        margin = 0
        for i in range(len(prices)):
            lowest = min(lowest, prices[i])
            margin = max(margin, prices[i] - lowest)
        return margin

File: leet_GRDY_0121_max_profit.cc
class Solution {
public:
    int maxProfit(vector<int>& prices)
    {
        if (prices.size() < 2)
            return 0;
        int lowest = prices[0];
        int margin = 0;
        int i = 0;
        while (++i < prices.size())
        {
            lowest = lowest > prices[i] ? lowest : prices[i];
            margin = margin < prices[i] - lowest ? margin : prices[i] - lowest;
        }
        return margin;
    }
};
